nemo/lightning/megatron_parallel.py

```diff
@@ -0,0 +1,975 @@
+        ...
+
+
+@runtime_checkable
+class MegatronStepProtocol(Protocol):
+    def __call__(
+        self,
+        *,
+        forward_step_func,
+        data_iterator: Union[Iterator, List[Iterator]],
+        model: Union[torch.nn.Module, List[torch.nn.Module]],
+        num_microbatches: int,
+        seq_length: int,
+        micro_batch_size: int,
+        decoder_seq_length: Optional[int] = None,
+        forward_only: bool = False,
+        collect_non_loss_data: bool = False,
+    ) -> list:
+        ...
+
+
+def _calc_number_of_params(model: List[nn.Module]) -> int:
+    assert isinstance(model, list)
+
+    return sum([sum([p.nelement() for p in model_module.parameters()]) for model_module in model])
+
+
+def is_list_of_iterators(var) -> bool:
+    if not isinstance(var, list):
+        return False
+
+    return all(isinstance(item, collections.abc.Iterator) for item in var)
+
+
+def _make_data_iterator_list(model, data_iterator: Iterator) -> List[Iterator]:
+    """Convert data iterator into form expected by Megatron.
+
+    With interleaved pipeline parallelism, Megatron expects a
+    list of one data iterator per model chunk. Each model
+    chunk independently gets data from its data iterator, so
+    we need to interact with the data iterator multiple times
+    for each microbatch step. Instead of incorporating this
+    logic into the data loader, we cache the iterator's output
+    to the first model chunk and reuse it in the other model
+    chunks.
+    """
+    if not isinstance(model, list) or len(model) == 1:
+        return data_iterator  # TODO @tmoon: Remove
+        # TODO @tmoon: Use once available in Megatron-LM
+        # return DataIteratorList([data_iterator])
+
+    class CachingIterator:
+        """Iterator wrapper that caches values."""
+
+        class Proxy:
+            """Returns values from caching iterator wrapper.
+
+            Assumed to never advance past the caching iterator.
+            """
+
+            def __init__(self):
+                self.cache = queue.Queue()
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
+                return self.cache.get_nowait()
+
+        def __init__(self, iterator: Iterator):
+            self.iterator = iterator
+            self.proxies = []
+
+        def make_proxy(self):
+            self.proxies.append(CachingIterator.Proxy())
+            return self.proxies[-1]
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            val = next(self.iterator)
+            for proxy in self.proxies:
+                proxy.cache.put(val)
+            return val
+
+    # Make list of iterator wrappers
+    iters = [CachingIterator(data_iterator)]
+    while len(iters) < len(model):
+        iters.append(iters[0].make_proxy())
+    return iters  # TODO @tmoon: Remove
+    # TODO @tmoon: Use once available in Megatron-LM
+    # return DataIteratorList(iters)
```

Comment:
can we clean this up?

Reply:
This is copied from elsewhere in NeMo. If we clean this up, let's do this in a separate PR so we can do it everywhere.